### ISSUE TITLE
Monorepos: multiple shards per repository

### ISSUE THREAD

**Issue by ysbaddaden** [crystal-lang/shards#635]
I'd like to start a discussion on monorepos. It is common for larger projects to maintain a number of shards or to split an overall framework into smaller pieces that can work individually as well as they can work together.

There are workarounds ([example](https://forum.crystal-lang.org/t/how-i-migrated-athena-to-a-monorepo-and-you-can-too/4154)) that accommodate the one shard per repo assumption of Shards by leveraging git submodules or subtrees, but they still need a repo for each shard and still need to synchronize from the main repo to each individual one.

Maybe shards could help?

**Comment by ysbaddaden** [comment]
For the installation side, maybe Shards could have a `root: sub/folder` argument to use as the root of a shard? The default would be to use the root of a shard.

For example:

```yaml
dependencies:
  framework:
    github: some/framework
```

Then the main shard for `framework`:
```yaml
dependencies:
  framework-router:
    github: some/framework
    root: shards/router

  framework-views:
    github: some/framework
    root: shards/views
```

That doesn't help with local and inter-shards dependencies, that would need to have `shard.override.yml`?

**Comment by ysbaddaden** [comment]
Another workaround is to have a single shard with different modules under `src`, treat them as if they were distinct shards (i.e. beware of requires), then have users only require the modules they need.

- src
  - src/router
  - src/views

Then:

```crystal
require "framework/router"
require "framework/views"
```

No need for special cross shards dependencies, pull a single repository, no need to fix local dependencies, a single `shard.yml` to handle for releases. The drawback is that you need to pull all the dependencies for the whole framework, which may be annoying & pointless if you only need a module that doesn't need them (damn).

**Comment by straight-shoota** [comment]
I understand an important reason for multiple shards is that they can evolve independently, with their own dependencies and releases. The monorepo development ensures individual components stay in sync, but they can move at different speeds.
If components are supposed to be usable independently, their versions should not be interlocked.

So the biggest challenge is probably version discovery. Shards pulls versions out of git tags, and those are per repository. This won't work for multiple subprojects with independent versions within one repository.

Namespaced version tags could be an option for that (tags `foo/bar:v1.1.0`, `foo/baz:v1.0.1`). I haven't put much thought about this, but it feels messy.

**Comment by straight-shoota** [comment]
I've had another idea in my head for a while now, which could potentially provide a solution for version discovery in a monorepo. But its main purpose is unrelated, so I'll have to write it down in a separate issue to keep this one focused.

**Comment by Blacksmoke16** [comment]
I can speak a bit on my experience with this.

> but they still need a repo for each shard and still need to synchronize from the main repo to each individual one.

I setup Terraform to manage the Athena GH org, so creating/configuring a new component repo is as simple as adding another entry to https://github.com/athena-framework/infrastructure/blob/master/terraform/github/components.tf, which uses a template repo as the source. So after applying the change, just have to go in and rename the types/files and update README and such.

Also added some tooling into the monorepo itself, so I can run a command and have it subtree a repo, then just have to add it to the main sync script and that's it. Takes some time to get it setup initially, but after that not _too_ bad adding new components and such.

> I understand an important reason for multiple shards is that they can evolve independently

Correct, this is why I went with the many-repo approach vs a monorepo where the user just has to do like `require "framework/router"` or `require "framework/shard"`. As I wanted each component to be versioned on its own, which needs its own repo anyway to tag releases and such.

Symfony on the other hand does things a bit differently by tagging all components with the same tag, skipping components that had no changes since the latest tag. Composer has https://getcomposer.org/doc/04-schema.md#replace, which allows the monorepo itself to be used as a dependency that satisfies all the specific components.

**Comment by paddor** [comment]
What's the status here? After implementing [ZMQ in pure Crystal](https://github.com/paddor/omq.cr), I decided to also give it an actual binding to [ZMQ in pure Rust](https://github.com/paddor/omq.rs) which belongs in that same repo, just like the other 5 bindings. But Crystal shards don't support this at all? I'm surprised, and not in a good way.

**Comment by ysbaddaden** [comment]
Nowhere.

I've been looking at some other package managers. Rust cargo and node NPM seem to have "workspaces" in the root's definition file, to tell embedded crates and packages.

I don't understand how ruby gem/bundler works, however. For example I can install `rails` from git and it will pickup all the nested gems, but none are referenced? Does bundler run a `glob("*.gemspec")`?

Maybe workspaces could be doable, somehow? For example:

```yaml
name: foo

workspaces:
  foo-client:
    path: bindings/crystal
```

Then for a dependency such as:

```yaml
dependencies:
  foo-client:
    git: https://bar/foo.git
```

The resolver would clone the `bar/foo` repository, load its top-level `shard.yml` for a ref, then instead of failing because the shard is `foo`, it would search the workspaces to load `bindings/crystal/shard.yml` that defines `foo-client` (found).

However, having different versions for each shard doesn't seem doable (or even needed). I wouldn't bother with it for the time being.

**Comment by paddor** [comment]
Yeah some kind of solution is needed. I could deal with a root level `shard.yml` but I don't want to put the Crystal binding into its own repo.

**Comment by straight-shoota** [comment]
Shard dependency references are already locations, not identifiers. So I don't think there's much point in having namespaces for resolution.
Also, it would be weird to have two different shards sharing the same reference:

```yaml
dependencies:
  foo-client:
    git: https://bar/foo.git
  foo-sdk:
    git: https://bar/foo.git
```

I don't see a reason why reference locations should not be able to specify subpaths within a repository, such as `git:https://bar/foo.git#bindings/crystal`. That's an instruction where to find the shard. It's clear and concise, avoiding ambiguity.

```yaml
dependencies:
  foo-client:
    git: https://bar/foo.git#bindings/crystal
  foo-sdk:
    git: https://bar/foo.git#bindings/crystal-sdk
```

**Comment by ysbaddaden** [comment]
We should read more about workspaces and why different package managers decided to go that route.

An advantage seems to be that the project and each library can depend on the libraries in the monorepo (regular dependency), then the package manager doesn't have to resolve and install the libs, it just uses the ones declared through the workspace.

With a root, be it a property or URI hash, shards would have to clone the repository and install the dependency, which defeats the point. You'd have to write, commit, and maintain an override for every shard in the monorepo.

**Comment by paddor** [comment]
OMQ.rs only needs subpath support: install shard from `bindings/crystal`.

Workspaces are for repos with multiple local packages that depend on each other. Package manager sees whole repo, links local deps, shares lockfile/build state, runs commands across packages.

Ruby/Bundler handles Rails by discovering .gemspec files in one git repo. That is closer to subpath/package discovery than full workspace behavior.

**Comment by straight-shoota** [comment]
@ysbaddaden Multiple shards within the same repo should be able to use `path` dependencies.

Plus, shards could be aware of shared repo location and subpaths when resolving dependencies with absolute URLs.

**Comment by ysbaddaden** [comment]
Not if you clone the repository or get a tarball then run `shards`.

**Comment by ysbaddaden** [comment]
I guess both approaches are valid for different use cases. There's value in a `root` (can put a shard anywhere in the repo), and there's value in workspaces (work with multiple, interdependent, small shards).

**Comment by paddor** [comment]
Status?

**Comment by ysbaddaden** [comment]
Somewhere at the bottom of our always increasing TODO.

No need to ask for update. No report == No update.
